cw_nano/lib/nano_wallet.dart

```diff
@@ -0,0 +1,440 @@
+    final String type = (_derivationType == DerivationType.nano) ? "standard" : "hd";
+
+    // our "mnemonic" is actually a seedkey:
+    if (!_mnemonic.contains(' ')) {
+      _seedKey = _mnemonic;
+    }
+
+    if (_seedKey == null) {
+      if (_derivationType == DerivationType.nano) {
+        _seedKey = bip39.mnemonicToEntropy(_mnemonic).toUpperCase();
+      } else {
+        _seedKey = await NanoUtil.hdMnemonicListToSeed(_mnemonic.split(' '));
+      }
+    }
+    _privateKey = await NanoUtil.uniSeedToPrivate(_seedKey!, 0, type);
+    _publicAddress = await NanoUtil.uniSeedToAddress(_seedKey!, 0, type);
+    this.walletInfo.address = _publicAddress!;
+
+    await walletAddresses.init();
+    await transactionHistory.init();
+    await save();
+  }
+
+  @override
+  int calculateEstimatedFee(TransactionPriority priority, int? amount) {
+    return 0; // always 0 :)
+  }
+
+  @override
+  Future<void> changePassword(String password) {
+    throw UnimplementedError("changePassword");
+  }
+
+  @override
+  void close() {
+    _client.stop();
+  }
+
+  @action
+  @override
+  Future<void> connectToNode({required Node node}) async {
+    try {
+      syncStatus = ConnectingSyncStatus();
+      final isConnected = _client.connect(node);
+      if (!isConnected) {
+        throw Exception("Nano Node connection failed");
+      }
+
+      try {
+        await _updateBalance();
+        await _updateRep();
+        await _receiveAll();
+      } catch (e) {
+        print(e);
+      }
+
+      syncStatus = ConnectedSyncStatus();
+    } catch (e) {
+      print(e);
+      syncStatus = FailedSyncStatus();
+    }
+  }
+
+  @override
+  Future<void> connectToPowNode({required Node node}) async {
+    _client.connectPow(node);
+  }
+
+  @override
+  Future<PendingTransaction> createTransaction(Object credentials) async {
+    credentials = credentials as NanoTransactionCredentials;
+
+    BigInt runningAmount = BigInt.zero;
+    await _updateBalance();
+    BigInt runningBalance = balance[currency]?.currentBalance ?? BigInt.zero;
+
+    final List<Map<String, String>> blocks = [];
+    String? previousHash;
+
+    for (var txOut in credentials.outputs) {
+      late BigInt amt;
+      if (txOut.sendAll) {
+        amt = balance[currency]?.currentBalance ?? BigInt.zero;
+      } else {
+        amt = BigInt.tryParse(
+                NanoUtil.getAmountAsRaw(txOut.cryptoAmount ?? "0", NanoUtil.rawPerNano)) ??
+            BigInt.zero;
+      }
+
+      if (balance[currency]?.currentBalance != null && amt > balance[currency]!.currentBalance) {
+        throw Exception("Trying to send more than entire balance!");
+      }
+
+      runningBalance = runningBalance - amt;
+
+      final block = await _client.constructSendBlock(
+        amountRaw: amt.toString(),
+        destinationAddress: txOut.extractedAddress ?? txOut.address,
+        privateKey: _privateKey!,
+        balanceAfterTx: runningBalance,
+        previousHash: previousHash,
```

Comment:
this is intentional, previousHash refers to the previous block in the chain, and would be null on the first loop since we aren't constructing that block

it is super weird that this works though, this first block should have a valid previous field, but it doesn't

I think it gets filled in when the transaction submits because we don't have access to the node / getAccountInfo rpc here

I'll double check to make sure this works but pretty sure that's what's happening

Reply:
nvm it gets filled in immediately

TLDR: it's supposed to be null, but only on the first run, it's set after so that it's set for the subsequent runs
it's how we're able to submit multiple nano transactions at the same time from the same account, since we know the block hash of the unsubmitted blocks